### ISSUE TITLE
Helper method/property to return styles from a styled-component

### ISSUE THREAD

**Issue by richardwestenra** [styled-components/styled-components#1209]
I am building a site which includes some CMS-editable content (inserted with `dangerouslySetInnerHTML`). I'll need to style this content with global/child element selectors instead of styled-components. The headings in the editable areas are using the same styles as my hand-coded styled-components, so I want to reuse the same styles in both places. At present, I'm using a mixin pattern where I define my styles with the `css` helper, then include them in both locations, like so:
```es6
const h3 = css`
  font-size: 2rem;
`;

const H3 = styled.h3`
  ${ h3 }
`;

const CMS_Content = styled.article`
  h3 {
    ${ h3 }
  }
`;
```

However I'd love to just define the styled-component once, and then export its styles in the CMS context. I realised that I can do this with `💅.componentStyle.rules.join('')`:
```es6
const H3 = styled.h3`
  font-size: 2rem;
`;

const CMS_Content = styled.article`
  h3 {
    ${ H3.componentStyle.rules.join('') }
  }
`;
```
but this seems a bit hacky, and I'm concerned that it might not always be supported in future versions.

It would be great if there were an officially-sanctioned helper method/prop, that one could use to accomplish this same pattern. e.g. something like this?
```es6
const H3 = styled.h3`
  font-size: 2rem;
`;

const CMS_Content = styled.article`
  h3 {
    ${ H3.css } // or H3.getCSS() or H3.componentStyle.get() or whatever
  }
`;
```
I'm totally open to naming suggestions for this method, the name isn't important to me. However it would be great to have an officially-documented way of doing this, because this can't be an uncommon use-case?

Alternatively, is `💅.componentStyle.rules.join('')` an okay pattern? I'm nervous about using an undocumented component property that might only be intended for internal use, and which might change without notice at any time.

**Comment by oliverturner** [comment]
Is there a reason this wouldn't work?

```
const h3 = css`
  font-size: 2rem;
`

const H3 = styled.h3`
  ${h3}
`;

const CMS_Content = styled.article`
  h3 {
    ${h3}
  }
`;
```

**Comment by richardwestenra** [comment]
@oliverturner That's exactly what I'm doing in my first example, and it's what I have currently in my codebase, and it works okay. It just seems unnecessarily inelegant to me? I'd rather avoid creating more variables than I really need.

**TLDR: Example 1 < Example 2 < Example 3, IMHO** :P

**Comment by oliverturner** [comment]
I should have scrolled up... sorry for being an inadvertent Captain Obvious!

**Comment by oliverturner** [comment]
I rather like the `${ H3.css }`: nice and elegant and plenty useful in all kinds of external context scenarios 👍

**Comment by mxstbr** [comment]
`💅.componentStyle.rules.join('')` is definitely not a public API or intended to be used that way at all, but I'm open to discussing adding a proper API for this. /cc @philpl @geelen

**Comment by philpl** [comment]
@mxstbr I’m a little conflicted here. First of all I think that such an API would open the door to bad practices, but then again it could be beautiful for libraries that want to add some smarter behaviour.

While I think that a CM/S should never serve raw HTML and CSS but an abstract structure of the content, like the Medium API does, given the right constraint it’s a valid use-case.

I’m aware that some people are already using this pattern for different things, so it’d make sense to create a getter for it.

The big contra is that once we introduce preprocessing this will most certainly not work in its current form. But, under the right constraints this API could return the exact input that was given to create the styled component.

This would make it possible to use the input of a styled component for a mixin or another styled component. This specific case could very well be implemented and would also work for preprocessing, for which we’d then return the preprocessed CSS abstract format

**Comment by richardwestenra** [comment]
Thanks @philpl! I hear you about not encouraging bad practices, but I think there are better ways of discouraging them than disallowing the functionality. I would include a note in the documentation to encourage users to use styled-components wherever possible, and only to use this feature where necessary. Indeed, this is exactly how [injectGlobal](https://www.styled-components.com/docs/api#) is handled: The feature is provided and documented, but explicitly discouraged.

The Medium-style CMS pattern sounds great, and I agree with you about avoiding serving raw HTML/CSS, but often it can't be helped. I'm using [GatsbyJS with Contentful](https://github.com/gatsbyjs/gatsby/tree/master/examples/using-contentful), which is serving Markdown, which is loaded and converted to an HTML string with [Remark](http://remark.js.org/) when the static site is generated by Gatsby. I'm not sure it's possible to do this as proper styled-components without parsing the HTML to JSON, which seems unnecessarily cumbersome. I doubt this is an unusual use-case - if anything, it's probably more common than not, no?

btw, in addition to the usage example that I mentioned above, I reckon that this feature would also be useful for extending multiple components, e.g.
```es6
const Card = styled.div`
  background: white;
  box-shadow: 0 0 3px grey;
`;

const Column = styled.div`
  width: 600px;
`;

const Section = styled.section`
  padding: 30px 0;
`;

const CardColumnSection = Section.extend`
  ${ Card.css }
  ${ Column.css }
  font-family: "Comic Sans MS" !important;
`;
```

**Comment by philpl** [comment]
@richardwestenra I think it’s reasonable to provide a getter that returns the exact input that was used to create a mixin/StyledComponent. This definition would always work

Would that work for all cases?

cc @mxstbr: :+1::-1:?

**Comment by MattHsiung** [comment]
Hey there! 
I opened up an [issue](https://github.com/styled-components/styled-components/issues/1248) today that might be a possible solution for this. I'll paste it below:

### Came across a use case for this today:

```javascript
const LabelWithFieldset = compose(
  Label.withComponent,
  Text.withComponent,
)('fieldset');

const Fieldset = LabelWithFieldset.extend`
  border: none;
  margin: 0;
  padding: 0;
`;
```
(where `Label` and `Text` are styled components ...incase that wasn't obvious 😆)

### Anyone else see any benefit to having something like `styled.compose/inherit`?

Something like this maybe:
```javascript
import { compose } from 'styled-components';

const LabelWithFieldset = compose(
  Label,
  Text,
)('fieldset')`
  border: none;
  margin: 0;
  padding: 0;
`;
```

It seems like a perfect tool for my current workflow with styled-components but maybe I'm off my rocker..

### Currently using this in my repo:
```javascript
import { compose } from 'recompose';

export default (...styledComponents) =>
	compose(...styledComponents.map(component => component.withComponent));
```

### Thoughts?

**Comment by MattHsiung** [comment]
Additionally, I just wanted to add that me and my team have been loving all the flexibility we've been getting from 💅 since v2. `withComponent`, `extend` and `attrs` are great additions to the api and allow for taking advantage of functional techniques with styled-components.

Thought I'd give an example of compose using the code that posted farther up:

```javascript
const Card = styled.div`
  background: white;
  box-shadow: 0 0 3px grey;
`;

const Column = styled.div`
  width: 600px;
`;

const Section = styled.section`
  padding: 30px 0;
`;

const CardColumnSection = compose(
  Card,
  Column,
)(Section)`
  font-family: "Comic Sans MS" !important;
`;
```

**Comment by mxstbr** [comment]
I'd like something to that effect much more than exposing another magic property on styled components.

**Comment by morajabi** [comment]
Sorry, @mxstbr what do you mean? I can't understand it.

**Comment by mxstbr** [comment]
I'd much rather having something like `import { compose } from 'styled-components'` or even `import { compose } from 'styled-compose'` than a magic property like `${Component.css}` or something.

**Comment by morajabi** [comment]
@mxstbr I see. Totally agree. It's also less confusing and easier to pick up, apart from the other things.

**Comment by diondiondion** [comment]
> It's also less confusing and easier to pick up

I'd disagree here, I was pretty confused about what's happening there when I saw the code. I don't think it's all that obvious that the compose function only takes the _styles_ of the components you pass to it while discarding the underlying component. My first assumption was that it would actually compose these components as children, JSX-style.

I thought this:
```
styled.compose(
  Card,
  Column,
)(Section)
```
would somehow result in this:
```
<Section>
    <Card></Card>
    <Column></Column>
</Section>
```

The usage of "another magic property" like `${Component.css}` on the other hand would be perfectly consistent with the way mixins (created using the css function) are used. The compose function would be _yet_ another special way in which styled components can be created, on top of the existing 3 ways:

```
styled.div``
styled(Comp)``
Comp.extend``
```

(Which is already 2 ways too many IMO, but that's besides the point.)

Might just be me, but this proposal doesn't feel elegant to me at all, and I'd much prefer an explicit native way of doing this. `$(Component.css}` just makes sense to me.

**Comment by morajabi** [comment]
> The compose function would be yet another special way in which styled components can be created, on top of the existing 3 ways:

We don't have to pack this with the styled-components itself, I said it's better to have it in a separate package.

> My first assumption was that it would actually compose these components as children, JSX-style.

I think it's due to the use of `compose` for HoCs. Maybe we can rename it?

> and I'd much prefer an explicit native way of doing this. $(Component.css} just makes sense to me.

I agree here. It's good to do something *natively* IMO, but the problem is that `.css` part which is confusing, what about `Comp.getStyles()`? Just it's more readable and the purpose of it is clear.

**Comment by diondiondion** [comment]
> I agree here. It's good to do something natively IMO, but the problem is that .css part which is confusing, what about Comp.getStyles()? Just it's more readable and the purpose of it is clear.

Indeed, `.css` looks a bit too much like a css file name extension. `.getStyles` would be a good alternative.

**Comment by philpl** [comment]
@mxstbr I don’t think it’s necessary to introduce a new helper for something that can be done with mixins/partials, I.e. with the css helper

**Comment by morajabi** [comment]
@philpl You're right, but I think @mxstbr main point was a separate package for that (`styled-compose`), but as I said you're right.

**Comment by richardwestenra** [comment]
I agree with @diondiondion, that the HoC pattern is more confusing and less intuitive.

However for me, the biggest problem with the HoC pattern is that it doesn't provide enough flexibility to solve my original problem: I need a way to extract the styles from a component to use them as child styles, e.g.
```es6
const Comp = styled.article`
  h3 {
    ${ H3.styles }
  }
`;
```
Using a helper property would solve this problem as well as allowing users to extend multiple components, in a way that's more consistent with traditional Sass/Less extends/mixins.

I'm fully on board with renaming the property to something else. I like `Comp.getStyles()` or `Comp.styles`? Other suggestions welcome.

**Comment by geelen** [comment]
I don't love the idea of `composes` TBH, since I think shared styles should be _mixed in_ to every component that needs it, not pulled out of some base `Text` component.

That said, I _also_ think @richardwestenra's final point is the killer blow. If you want to pull styles off a component to put somewhere, it's best if you have total flexibility of where to put them.

Final point, though is `.rules` would be a better name, since that's what they are (they're only styles once they've all been evaluated against the current props & theme)

> import { compose } from 'styled-compose' than a magic property like ${Component.css} or something.

@mxstbr this isn't actually a magic property. If `Component.rules` returns a `RuleSet` (or an array, or a block, or a function, etc) it'll get interpolated with the exact same logic as every other interpolation. The magic is already there, let's just expose it!

**Comment by morajabi** [comment]
@geelen 

> Final point, though is .rules would be a better name, since that's what they are (they're only styles once they've all been evaluated against the current props & theme)

Thanks for the clarification and `rules` is not confusing and it's great.

> H3.componentStyle.rules.join('')

What is that `componentStyle` style?

**Comment by ryanzec** [comment]
So it does not seem like there has been much progress on this issue in a while however I was pointed to it from spectrum and would like to add my use case to it.

I also would benefit from having access to the styles of the component as a string, but my use case is for testing. I use AVA for testing and is does not support really any integration with styled-component like some other testing framework do but truth me told I don't need that much of an integration. If I need to be able to test a styled component is rendering JSX properly, I can use existing react testing tools. All I do need is some tooling for testing the css generated by styled components.

Now unit testing the styled components is not so much about making sure the values are right (since I can't read HEX colors very well :) ), it is more so about knowing what components will be effected when I change a global css variable / css utility function (aka. sass mixin). If I were able to get the css output of a component, I could just use AVA snapshot testing functionality (like a number of testing framework have) and it would be easy to know when styles change.

Right now I have a less than ideal solution for this in which my `styled*` calls define no styles directly but instead just call functions and those functions just return strings that define the styles, like:

```js
export const baseContainerStyles = (props) => {
  return `
    color: white;
    background-color: {props.backgroundColor};
  `;
};

export const MyComponent = styled.div`
  ${baseContainerStyles}
`;
```

With this I am able to test the functions that return the styles easily (after some simple data massaging) with snapshot as it is just a string.

Now I actually like having the function call however I lose some editor / IDE support as I get no highlighting / intellisense. Now since the `css` utility just returns an array of strings right now, I think I might use that in my functions defining the styles however the result of the `css` utility is internal and might change since it is not designed to be used outside of styled itself.

Having an official API (and how it is exposed is not something I am too worried about) for accessing the generated styled of a component as a string purely on the client side would enable tooling like this without having to go through the different things that it seems `jest-styled-components` has to do.

As a note, I did look at `jest-styled-components` and tried to do the same things it does but was meet with a very brittle system that did not do everything I needed and kept breaking something when fixing something. I am sure with enough time I could get something working but I don't see that as a great investment of time when all I am really looking for is the generated css.

**Comment by kitten** [comment]
@ryanzec accessing styles from a component will never be possible as they're not deterministic and constant depending on the input. That's why I suggested to output the input or processed input, whatever that might be once we consider this.

I'd really love to have some testing utility for Ava, but please reach out to me :wink: You will have to use the StyleSheet unfortunately to observe the *output* of those components, and the rules that they eventually mount.

I was still looking at making this API nicer, especially for jest-styled-components, but another perspective would help. But this issue is kind of unrelated, so let's keep this to an issue, if you prefer it, and call it "ava-styled-components" or something 🙌

**Comment by ryanzec** [comment]
@kitten Sorry I was not clear, I was assuming the API to access the styles would have to have an input like `StyledComponent.generateStyles(props)`.

Maybe I am missing something but what is the benefit in adding the complexity of the StyleSheet, mounting / unmounting components, parsing out the styles from what happens during mounting, etc vs:

```js
// testUtils.js
export const cleanCss = (cssString) => {
  const results = sass.renderSync({
    data: `.test{${cssString}}`,
    outputStyle: 'expanded',
  });

  return `\n${results.css.toString()}`;
};

// component.spec.js
test('baseContainerStyles generates css properly', (t) => {
  t.snapshot(testUtils.cleanCss(component.baseContainerStyles({color: 'red'}));
});
```

**Comment by kitten** [comment]
@ryanzec even that is technically separate and while this would certainly be possible to what end would it make sense? The idea of some test utilities would certainly be to observe the actual styles that will end up in the stylesheet, so isn't it preferable to be closer to the actual pipeline?

Anyhow, I'm very much fond of the idea of having the ability to take the input of one StyledComponent and create another with it for even higher composability, but what is it you are trying to achieve?

We have jest-styled-component, so we have in fact proof of the feasibility of the current testing approach, so what speaks against that?

In your example I'm also seeing that you're testing a single component with hypothetical props and some functions around it. Does this make sense? Certainly it'd be closer to what you'd expect to be able to write JSX for that component or another, or multiple components and observe all their styles in a snapshot.

**Comment by ryanzec** [comment]
@kitten I guess my view point on what should be tested in the unit tests that users write that are using `styled-components` is a bit different.

To answer your question about testing the component itself first, this is how I do that (with the css testing too):

```js
test('render properly with passed in props', (tester) => {
  const props = {
    isThin: true,
    isPill: true,
    styleType: 'default',
  };
  const wrapper = shallow(<Button {...props}>button</Button>);

  tester.snapshot(testUtilities.cleanCss(component.baseContainerStyles(props)));
  tester.snapshot(toJson(wrapper));
});
```

Since `Button` is just a `styled.div` which just returns a react component, I should not have to do anything special to test it outside of what I normally do to test a react component.

Like I said before, I am sure it is feasibility to do the level of testing that `jest-styled-components` does with AVA (though having as deep an integration might not be since I think extending AVA is not as easy last I read) however for me is comes down to simple ROI. In the about 3 hours I spent trying to deconstruct `jest-styled-components`, I ended up with a system that somewhat worked (there was some bug I forget when I finally gave up) but it had a bunch of boilerplate code with the tests. The system that I have in place now took about 15 minutes and it does exactly what it needs to ensure my code is correct. The API to `styled-components` is passing a string literal of styles and I am able to verify that string is correct. Now if `styled-components` is not doing the right things with that string:

1. That means there is a bug in `styled-component` and not my code (and if my unit test is surfacing that, then it is no longer a unit test)
2. This is something that will come out during manual / regression / integration testing (unit testing does not replace this, it is just a different layer of testing in my opinion).

I try to follow this patterns for all my testing when it makes sense but just like with almost everything I do, no rule should be followed 100% of the time. Testing react rendering is 1 exception to this rules as without invoking some of the react internals, testing the rendering of component would be so much harder. However when it comes to react life cycle methods, I don't actually test them by invoking them through react. I test the method itself outside of react because I have a level of expectation that react will do the right thing (and I find out quick if it is not in my manual testing).

`styled-components` does not fall under this expect to me. All the APIs I have seen (`styled.ELEMENT`, `styled(COMPONENT)`, `COMPONENT.extend`, `css`, etc.) just take in a string for styles so in my opinion that is all that is needs to be tested (if there were testing utils that did more, I would probably used them, they are just not worth the time investment from my vantage point).

**Comment by jpvalery** [comment]
@richardwestenra what did you end up doing for this issue? 

I'm currently hacking something with Gatsby + Contentful and I'm trying to display a p with a different color than the global p without success

`const Desc = styled.p`
 color: #000000 !important;
 ``
followed by

`  <Desc
        dangerouslySetInnerHTML={{
          __html: props.desc.childMarkdownRemark.html,
        }}
        />`

But I couldn't find any solution (or was asking the question the wrong way). 

If anybody has an idea of what I'm doing wrong or a better fix than encoding <span> in Contenful, I'm all ears :)

**Comment by richardwestenra** [comment]
@jpvalery I ended up going with the [Example 1 approach from my original comment](#issue-263146426), for want of a better option. For many of my most-used styles (e.g. button styles), I've got a `const button = styled.css` object, which is then included in a `const Button = styled.Button` component, and is also included as a child of other SCs (e.g. my Gatsby Contentful CMS areas:
```
const CMS_Content = styled.article`
  button {
    ${ button }
  }
`;
```

I tend to think of them as like having both a Sass mixin and some somewhat-extendable Sass classes. It's not super ideal but it works 🤷‍♂️

**Comment by danielfdsilva** [comment]
This would definitely be a very useful feature.
I bring yet another use case that I wasn't able to solve using @richardwestenra suggestion. 

Imagine we have a button configurable through props:

```css
const buttonStyles = css`
  /* Adapt the colors based on primary prop */
  background: ${props => props.primary ? "palevioletred" : "white"};
  color: ${props => props.primary ? "white" : "palevioletred"};

  font-size: 1rem;
`;

const Button = styled.button`
  ${buttonStyles}
`;
```

and I need to apply the styles to a button of a third party plugin. I'd need to use the `createGlobalStyle` to set the styles for the button's class, but I'd also need a way to pass the attributes.
Using the suggested `Comp.getStyles()` we could pass props similar to `.attr()`.

```css
.third-party-button {
  ${Button.getStyles(() => ({ primary: true })}
}
```

Alternatively, is there a way to pass static props to `css`?
Anyone has ideas for this use case?

**Comment by mqklin** [comment]
I use this: `C.componentStyle.rules.filter(r => typeof r === 'string').join('')`

**Comment by lifeiscontent** [comment]
@mxstbr is there a function in styled-components to parse the css template literal as an object? I'm building a UI Library and this would be really helpful as a utility.

**Comment by CameronJ26** [comment]
I've got a similar problem. I'm importing the PlaidLink component from react-plaid-link. That component outputs HTML like this:
```
<div>
  <button />
</div>
```
I already have a Styled Component called `StyledButton`, and I want to apply the same style to the PlaidLink button. If I could get the css from `StyledButton` I could pass it to Plaid. Otherwise it's unclear how to support this.

**Comment by topolanekmartin** [comment]
Is any progress on this issue? I am looking for something similar as @richardwestenra mentioned. 

I keep everything DRY if it is possible. And I feel bad every time I need to repeat code for a Button that I want to apply on Link, for example. To avoid duplicating code, I can create a variable and reuse the variable instead. But here is the thing. I could create many and many variables (and their various states) for all atomic components which I don't want to. 

There is another use case. Sometimes I don't want styles of the whole component but only hover state. So it would be nice to get only hover styles. E. g.:

```
const buttonStyles = StyledButton.getStyles();
const buttonHoverStyles = StyledButton.getStyles('hover').attrs({ type: 'primary' });
```

Or something like that.

Would it be achievable? 

Thanks in advance!

**Comment by kitten** [comment]
I think while we did mull this over a lot and it has come up and been bumped every now and again, we've really struggled to find a good reason to do this and a good API that'd make sense for this. Without specific use-cases that fit into what styled-components is and does — currently — we'd risk adding an API that wouldn't make a lot of sense.
